tensorcircuit/templates/lattice.py

```diff
@@ -0,0 +1,1526 @@
+    ):
+        """Initializes the CustomizeLattice."""
+        super().__init__(dimensionality)
+        if len(identifiers) != len(coordinates):
+            raise ValueError(
+                "Identifiers and coordinates lists must have the same length."
+            )
+
+        # The _build_lattice logic is simple enough to be in __init__
+        self._identifiers = list(identifiers)
+        self._coordinates = [np.array(c) for c in coordinates]
+        self._indices = list(range(len(identifiers)))
+        self._ident_to_idx = {ident: idx for idx, ident in enumerate(identifiers)}
+
+        # Validate coordinate dimensions
+        for i, coord in enumerate(self._coordinates):
+            if coord.shape != (dimensionality,):
+                raise ValueError(
+                    f"Coordinate at index {i} has shape {coord.shape}, "
+                    f"expected ({dimensionality},)"
+                )
+
+        logger.info(f"CustomizeLattice with {self.num_sites} sites created.")
+
+        if precompute_neighbors is not None and precompute_neighbors > 0:
+            self._build_neighbors(max_k=precompute_neighbors)
+
+    def _build_lattice(self, *args: Any, **kwargs: Any) -> None:
+        """For CustomizeLattice, lattice data is built during __init__."""
+        pass
+
+    def _build_neighbors(self, max_k: int = 1, **kwargs: Any) -> None:
+        """Calculates neighbors using a KDTree and caches the distance matrix.
+
+        This method uses a memory-efficient approach to identify neighbors. It
+        first leverages `scipy.spatial.distance.pdist` to compute all pairwise
+        distances and identify the unique neighbor shells. It then uses a
+        `scipy.spatial.KDTree` for fast radius queries to find the neighbors
+        in each shell.
+
+        Crucially, after finding the neighbors, it reuses the initial `pdist`
+        result to construct and cache the full N x N distance matrix in
+        `self._distance_matrix`, avoiding any re-computation if the matrix is
+        accessed later.
+
+        :param max_k: The maximum number of neighbor shells to
+            calculate. Defaults to 1.
+        :type max_k: int, optional
+        :param tol: The numerical tolerance for distance
+            comparisons. Defaults to 1e-6.
+        :type tol: float, optional
+        """
+        if self.num_sites < 2:
+            return
+
+        tol = kwargs.get("tol", 1e-6)
+
+        all_coords = np.array(self._coordinates)
+        all_distances_sq = pdist(all_coords, metric="sqeuclidean")
+        dist_shells_sq = self._identify_distance_shells(all_distances_sq, max_k, tol)
+
+        if not dist_shells_sq:
+            logger.info("No distinct neighbor shells found.")
+            return
+
+        tree = KDTree(all_coords)
+        self._neighbor_maps = {}
+
+        found_neighbors_so_far: List[set[int]] = [set() for _ in range(self.num_sites)]
+
+        for k_idx, target_d_sq in enumerate(dist_shells_sq):
```

Comment:
why the neigbor building logic here is also modified? The original implementation is more neat? what is the gain here?